### PR TITLE
Optimize translator visitors

### PR DESCRIPTION
**Pull request by SergeiPavlov**
* Refactor
* avoid double dereference of the same properties
* Use arrays instead of List<> when number of elements is known.

### CODE REVIEW

File: Orm/Xtensive.Orm/Orm/Providers/Expressions/ExpressionProcessor.Helpers.cs [hunk 1]
```diff
@@ -50,8 +50,7 @@ private SqlExpression TryTranslateCompareExpression(BinaryExpression expression)
       var methodType = method.DeclaringType;
 
       // There no methods in IComparable except CompareTo so checking only DeclatingType.
-      bool isCompareTo = methodType==WellKnownInterfaces.Comparable
-        || (methodType.IsGenericType && methodType.CachedGetGenericTypeDefinition() == WellKnownInterfaces.ComparableOfT);
+      bool isCompareTo = methodType==WellKnownInterfaces.Comparable || methodType.IsGenericType(WellKnownInterfaces.ComparableOfT);
```
Comment (botinko) [inline Orm/Xtensive.Orm/Orm/Providers/Expressions/ExpressionProcessor.Helpers.cs:53]:
```suggestion
      bool isCompareTo = methodType is WellKnownInterfaces.Comparable || methodType.IsGenericType(WellKnownInterfaces.ComparableOfT);
```

File: Orm/Xtensive.Orm.Tests/Linq/MemberCompilerProviderTest.cs [hunk 2]
```diff
@@ -147,7 +147,7 @@ public void ArrayAndEnumerableOverloadTest()
       var eachEnumerable = typeof (EachExtensions)
         .GetMethods()
         .Single(method => method.GetParameterTypes()
-          .Any(type => type.IsGenericType && type.CachedGetGenericTypeDefinition()==typeof(IEnumerable<>)))
+          .Any(type => type.IsGenericType(typeof(IEnumerable<>))))
```
Comment (botinko) [inline Orm/Xtensive.Orm.Tests/Linq/MemberCompilerProviderTest.cs]: `typeof(IEnumerable<>)` may be WellKnown one.

File: Orm/Xtensive.Orm/Linq/ExpressionVisitor.cs [hunk 3]
```diff
@@ -253,23 +228,34 @@ protected virtual MemberMemberBinding VisitMemberMemberBinding(MemberMemberBindi
...
-        var originalBinding = original[i];
-        MemberBinding p = VisitBinding(originalBinding);
-        results.Add(p);
-        isChanged |= !ReferenceEquals(originalBinding, p);
+        var originalValue = original[i];
+        var p = func(originalValue);
+        if (ar != null) {
+          ar[i] = p;
+        }
+        else if (!ReferenceEquals(p, originalValue)) {
```
Comment (botinko) [inline Orm/Xtensive.Orm/Linq/ExpressionVisitor.cs:243]: nice optimization =)

**Review by botinko — APPROVED** [review]

File: Orm/Xtensive.Orm.Tests/Linq/MemberCompilerProviderTest.cs (hunk 2 above)
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm.Tests/Linq/MemberCompilerProviderTest.cs]: done

File: Orm/Xtensive.Orm/Orm/Providers/Expressions/ExpressionProcessor.Helpers.cs (hunk 1 above)
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Orm/Providers/Expressions/ExpressionProcessor.Helpers.cs:53]: It is syntactically incorrect.    `is` operator is not the same as `==`

**Pull request merged**